File: OpenHarmony/developtools/hapsigner/hapsigntool_cpp/utils/src/file_utils.cpp
/*
 * Copyright (c) 2024-2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include <iostream>
#include <filesystem>
#include <fstream>
#include <climits>

#include "string_utils.h"
#include "signature_tools_errno.h"
#include "file_utils.h"

namespace OHOS {
namespace SignatureTools {
namespace fs = std::filesystem;

const int FileUtils::NUM_TWO = 2;
const int FileUtils::NUM_THREE = 3;
const int FileUtils::NUM_FOUR = 4;
const std::string FileUtils::LIBS_PATH_PREFIX = "libs/";
bool FileUtils::IsEmpty(std::string cs)
{
    if (cs.length() == 0) {
        return true;
    }
    return false;
}

bool FileUtils::IsSpaceEnough(const std::string& filePath, const int64_t requiredSpace)
{
    uint64_t freeSpace = 0;
    struct statfs diskStatfs;
    int ret = statfs(filePath.c_str(), &diskStatfs);
    if (ret >= 0) {
        freeSpace = (uint64_t)diskStatfs.f_bsize * (uint64_t)diskStatfs.f_bavail;
    } else {
        SIGNATURE_TOOLS_LOGE("statfs fail, error code = %d", ret);
    }
    return freeSpace >= static_cast<uint64_t>(requiredSpace);
}

std::string FileUtils::GetSuffix(std::string filePath)
{
    if (filePath.empty()) {
        return "";
    }
    size_t lastDotPosition = filePath.rfind(".");
    bool positionFlag = (lastDotPosition == std::string::npos) || (lastDotPosition == filePath.size() - 1);
    if (positionFlag) {
        return "";
    }
    return filePath.substr(lastDotPosition + 1);
}

bool FileUtils::ValidFileType(const std::string& filePath, const std::initializer_list<std::string> types)
{
    std::string suffix = GetSuffix(filePath);
    bool flag = suffix.empty() || (StringUtils::ContainsCase(types, suffix) == false);
    if (flag) {
        PrintErrorNumberMsg("NOT_SUPPORT_ERROR", NOT_SUPPORT_ERROR, "Not support file: " + filePath);
        return false;
    }
    return true;
}

int FileUtils::Write(const std::string& content, const std::string& output)
{
    std::ofstream outFile(output, std::ios::binary);
    bool flag = (outFile.rdstate() != 0);
    if (flag) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "open output file: " + output + " failed");
        return IO_ERROR;
    }
    outFile.write(&content[0], content.size());
    if (outFile.rdstate() != 0) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "Failed to write data to output stream.");
        return IO_ERROR;
    }
    return RET_OK;
}

int FileUtils::Read(std::ifstream& input, std::string& ret)
{
    ret.clear();
    if (input.rdstate() != 0) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "invalid input stream");
        return IO_ERROR;
    }
    ret.clear();
    std::string buffer(FileUtils::FILE_BUFFER_BLOCK, 0);
    while (input) {
        input.read(&buffer[0], buffer.size());
        if (input.fail() && !input.eof()) {
            SIGNATURE_TOOLS_LOGE("error occurred while reading data");
            return IO_ERROR;
        }
        ret.append(&buffer[0], input.gcount());
    }
    return RET_OK;
}

int FileUtils::ReadFile(const std::string& path, std::string& ret)
{
    std::ifstream file(path, std::ios::binary);
    bool flag = (file.rdstate() != 0);
    if (flag) {
        PrintErrorNumberMsg("IO_ERROR", IO_ERROR, "open input file: " + path + " failed");
        return IO_ERROR;
    }
    if (Read(file, ret) < 0) {
        SIGNATURE_TOOLS_LOGE("read error!");
        return IO_ERROR;
    }
    return RET_OK;
}

int FileUtils::ReadFileByOffsetAndLength(std::ifstream& file, size_t offset, size_t length, std::string& ret)
{
    if (length > INT_MAX) {
        SIGNATURE_TOOLS_LOGE("Size cannot be greater than Integer max value: %zu", length);
        return RET_FAILED;
    }
    if (ReadInputByOffsetAndLength(file, offset, length, ret) < 0) {
        SIGNATURE_TOOLS_LOGE("Error readInputByOffsetAndLength");
        return IO_ERROR;
    }
    return RET_OK;
}

int FileUtils::ReadInputByOffsetAndLength(std::ifstream& input, size_t offset, size_t length, std::string& ret)
{
    if (length > INT_MAX) {
        SIGNATURE_TOOLS_LOGE("Size cannot be greater than Integer max value: %zu", length);
        return -1;
    }
    input.seekg(offset);
    if (ReadInputByLength(input, length, ret) < 0) {
        SIGNATURE_TOOLS_LOGE("Error readInputByLength");
        return -1;
    }
    return RET_OK;
}

int FileUtils::ReadInputByLength(std::ifstream& input, size_t length, std::string& ret)
{
    if (length > INT_MAX) {
        SIGNATURE_TOOLS_LOGE("Size cannot be greater than Integer max value: %zu", length);
        return -1;
    }
    if (input.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Error input");
        return -1;
    }
    ret.clear();

    char* buffer = new (std::nothrow)char[FILE_BUFFER_BLOCK];
    if (buffer == NULL) {
        SIGNATURE_TOOLS_LOGE("create buffer error!");
        return -1;
    }
    size_t hasReadLen = 0;

    while (hasReadLen < length && input) {
        int readLen = static_cast<int>(std::min(length - hasReadLen, (size_t)FILE_BUFFER_BLOCK));
        input.read(buffer, readLen);
        bool flag = (input.gcount() != readLen);
        if (flag) {
            delete[] buffer;
            SIGNATURE_TOOLS_LOGE("read %zu bytes data less than %zu", hasReadLen, length);
            return -1;
        }
        ret.append(buffer, readLen);
        hasReadLen += readLen;
    }
    delete[] buffer;
    if (hasReadLen != length) {
        SIGNATURE_TOOLS_LOGE("read %zu bytes data less than %zu", hasReadLen, length);
        return -1;
    }
    return RET_OK;
}

bool FileUtils::AppendWriteFileByOffsetToFile(std::ifstream& input, std::ofstream& out, size_t offset, size_t size)
{
    if (input.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("input failed.");
        return false;
    }
    if (out.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Failed get out stream");
        return false;
    }
    input.seekg(offset);
    if (WriteInputToOutPut(input, out, size) < 0) {
        SIGNATURE_TOOLS_LOGE("Error: writeInputToOutPut");
        return false;
    }
    return true;
}

bool FileUtils::AppendWriteFileToFile(const std::string& inputFile, const std::string& outputFile)
{
    std::ifstream input(inputFile, std::ios::binary);
    std::ofstream output(outputFile, std::ios::binary | std::ios::app);
    bool flag = (0 != input.rdstate());
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to get input stream object!");
        return false;
    }
    flag = (0 != output.rdstate());
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to get output stream object!");
        return false;
    }
    char* buffer = new char[FILE_BUFFER_BLOCK];
    while (!input.eof()) {
        input.read(buffer, FILE_BUFFER_BLOCK);

        if (input.fail() && !input.eof()) {
            SIGNATURE_TOOLS_LOGE("error occurred while reading data");
            delete[]buffer;
            return false;
        }
        std::streamsize readLen = input.gcount();
        if (readLen > 0) {
            output.write(buffer, readLen);
        }
        if (!output) {
            SIGNATURE_TOOLS_LOGE("error occurred while writing data");
            delete[]buffer;
            return false;
        }
    }
    delete[]buffer;
    return true;
}

bool FileUtils::AppendWriteByteToFile(const std::string& bytes, const std::string& outputFile)
{
    std::ofstream output(outputFile, std::ios::binary | std::ios::app);
    bool flag = (WriteByteToOutFile(bytes, output) == false);
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to write data to output stream, outfile: %s", outputFile.c_str());
        return false;
    }
    return true;
}

int FileUtils::WriteInputToOutPut(std::ifstream& input, std::ofstream& output, size_t length)
{
    int result = RET_OK;
    char* buf = new (std::nothrow)char[FILE_BUFFER_BLOCK];
    if (buf == NULL) {
        SIGNATURE_TOOLS_LOGE("create buffer error!");
        return RET_FAILED;
    }

    while (input) {
        int min = std::min(static_cast<int>(length), FILE_BUFFER_BLOCK);
        input.read(buf, min);
        if (input.fail() && !input.eof()) {
            SIGNATURE_TOOLS_LOGE("read error!");
            delete[] buf;
            return IO_ERROR;
        }
        length -= input.gcount();
        output.write(buf, input.gcount());
        if (!output.good()) {
            SIGNATURE_TOOLS_LOGE("write error!");
            delete[] buf;
            return IO_ERROR;
        }
        
        if (length <= 0) {
            break;
        }
    }
    delete[] buf;
    // After the file is written, datasize must be 0, so the if condition will never hold
    if (length != 0) {
        SIGNATURE_TOOLS_LOGE("written length error!");
        return IO_ERROR;
    }
    return result;
}

bool FileUtils::WriteInputToOutPut(const std::string& input, const std::string& output)
{
    std::ifstream in(input, std::ios::binary);
    std::ofstream out(output, std::ios::binary);
    bool flag = (in.rdstate() != 0);
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to get input stream object!");
        return false;
    }
    flag = (out.rdstate() != 0);
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to get output stream object!");
        return false;
    }
    char* buffer = new char[FILE_BUFFER_BLOCK];
    while (!in.eof()) {
        in.read(buffer, FILE_BUFFER_BLOCK);

        if (in.fail() && !in.eof()) {
            SIGNATURE_TOOLS_LOGE("error occurred while reading data");
            delete[]buffer;
            return false;
        }

        std::streamsize readLen = in.gcount();
        if (readLen > 0) {
            out.write(buffer, readLen);
        }

        if (!out) {
            SIGNATURE_TOOLS_LOGE("error occurred while writing data");
            delete[]buffer;
            return false;
        }
    }
    delete[]buffer;
    return true;
}

bool FileUtils::WriteByteToOutFile(const std::string& bytes, const std::string& outFile)
{
    std::ofstream ops(outFile, std::ios::binary);
    bool flag = (WriteByteToOutFile(bytes, ops) == false);
    if (flag) {
        SIGNATURE_TOOLS_LOGE("Failed to write data to ops, outfile: %s", outFile.c_str());
        return false;
    }
    return true;
}

bool FileUtils::WriteByteToOutFile(const std::string& bytes, std::ofstream& outFile)
{
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Failed to get output stream object, outfile");
        return false;
    }
    outFile.write(&bytes[0], bytes.size());
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Failed to write data to ops, outfile ");
        return false;
    }
    outFile.flush();
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Flush error");
        return false;
    }
    return true;
}

bool FileUtils::WriteByteToOutFile(const std::vector<int8_t>& bytes, std::ofstream& outFile)
{
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Failed to get output stream object, outfile");
        return false;
    }
    outFile.write((char*)&bytes[0], bytes.size());
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Failed to write data to ops, outfile ");
        return false;
    }
    outFile.flush();
    if (outFile.rdstate() != 0) {
        SIGNATURE_TOOLS_LOGE("Flush error");
        return false;
    }
    return true;
}

bool FileUtils::IsRunnableFile(const std::string& name)
{
    if (name.empty()) {
        return false;
    }
    size_t dotPos = name.rfind('.');
    if (dotPos == std::string::npos) {
        return false;
    }
    std::string suffix = name.substr(dotPos + 1);
    if (suffix == "an" || suffix == "abc") {
        return true;
    }
    std::string libDir = name.substr(0, LIBS_PATH_PREFIX.size());
    if (LIBS_PATH_PREFIX.compare(libDir) == 0) {
        return true;
    }
    return false;
}

bool FileUtils::IsValidFile(std::string file)
{
    std::filesystem::path filePath = file;
    bool flag = std::filesystem::exists(filePath);
    if (!flag) {
        PrintErrorNumberMsg("FILE_NOT_FOUND", FILE_NOT_FOUND, "'" + file + "' file is not exists");
        return false;
    }
    flag = std::filesystem::is_directory(filePath);
    if (flag) {
        PrintErrorNumberMsg("FILE_NOT_FOUND", FILE_NOT_FOUND, "'" + file + "' file is a directory not file");
        return false;
    }
    return true;
}

int64_t FileUtils::GetFileLen(const std::string& file)
{
    std::filesystem::path filePath = file;
    bool flag = std::filesystem::exists(filePath) && std::filesystem::is_regular_file(filePath);
    if (flag) {
        return std::filesystem::file_size(filePath);
    }
    return -1;
}

void FileUtils::DelDir(const std::string& file)
{
    std::filesystem::path filePath = file;
    bool flag = std::filesystem::is_directory(filePath);
    if (flag) {
        for (auto& p : std::filesystem::recursive_directory_iterator(filePath)) {
            DelDir(p.path());
        }
    }
    std::filesystem::remove(file);
    return;
}

} // namespace SignatureTools
} // namespace OHOS

File: OpenHarmony/developtools/hapsigner/hapsigntool/hap_sign_tool_lib/src/main/java/com/ohos/hapsigntool/utils/FileUtils.java
/*
 * Copyright (c) 2021-2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ohos.hapsigntool.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.ohos.hapsigntool.error.CustomException;
import com.ohos.hapsigntool.error.ERROR;
import com.ohos.hapsigntool.error.SignToolErrMsg;

import java.io.BufferedWriter;
import java.io.ByteArrayOutputStream;
import java.io.Closeable;
import java.io.DataOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.nio.charset.Charset;
import java.nio.file.Files;

/**
 * Common file operation.
 *
 * @since 2021/12/28
 */
public final class FileUtils {
    /**
     * LOGGER.
     */
    private static final LogUtils LOGGER = new LogUtils(FileUtils.class);

    /**
     * add GSON static.
     */
    public static final Gson GSON = (new GsonBuilder()).disableHtmlEscaping().create();

    /**
     * add GSON_PRETTY_PRINT static.
     */
    public static final Gson GSON_PRETTY_PRINT = (new GsonBuilder()).disableHtmlEscaping().setPrettyPrinting().create();

    /**
     * File reader block size
     */
    public static final int FILE_BUFFER_BLOCK = 1024 * 1024;

    /**
     * File end
     */
    public static final int FILE_END = -1;

    /**
     * Expected split string length
     */
    public static final int SPLIT_LENGTH = 2;

    /**
     * libs dir path
     */
    public static final String LIBS_PATH_PREFIX = "libs/";

    /**
     * abc file suffix
     */
    public static final String ABC_FILE_SUFFIX = ".abc";

    /**
     * an file suffix
     */
    public static final String NATIVE_LIB_AN_SUFFIX = ".an";

    /**
     * bitmap file name
     */
    public static final String BIT_MAP_FILENAME = ".pages.info";

    private static final byte[] EMPTY_BYTE_ARRAY = new byte[0];

    private FileUtils() {
    }

    /**
     * Close closeable quietly.
     *
     * @param closeable closeable
     */
    public static void close(Closeable closeable) {
        if (closeable != null) {
            try {
                closeable.close();
            } catch (IOException exception) {
                LOGGER.debug(exception.getMessage(), exception);
            }
        }
    }

    /**
     * Read byte from input file.
     *
     * @param file Which file to read
     * @return byte content
     * @throws IOException Read failed
     */
    public static byte[] readFile(File file) throws IOException {
        return read(Files.newInputStream(file.toPath()));
    }

    /**
     * Read byte from input stream.
     *
     * @param input Input stream
     * @return File content
     * @throws IOException Read failed
     */
    public static byte[] read(InputStream input) throws IOException {
        try (ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            byte[] buffer = new byte[FILE_BUFFER_BLOCK];
            int read;
            while ((read = input.read(buffer)) != FILE_END) {
                output.write(buffer, 0, read);
            }
            return output.toByteArray();
        } finally {
            close(input);
        }
    }

    /**
     * Read byte from input file.
     *
     * @param file input file
     * @param offset offset
     * @param length length
     * @return data bytes
     */
    public static byte[] readFileByOffsetAndLength(File file, long offset, long length) throws IOException {
        try (FileInputStream input = new FileInputStream(file)) {
            return readInputByOffsetAndLength(input, offset, length);
        }
    }

    /**
     * Read byte from input stream.
     *
     * @param input input stream
     * @param offset offset
     * @param length length
     * @return data bytes
     * @throws IOException read exception
     */
    public static byte[] readInputByOffsetAndLength(InputStream input, long offset, long length) throws IOException {
        input.skip(offset);
        return readInputByLength(input, length);
    }

    /**
     * Read byte from input stream.
     *
     * @param input InputStream
     * @param length length
     * @return data bytes
     */
    public static byte[] readInputByLength(InputStream input, long length) throws IOException {
        try (ByteArrayOutputStream output = new ByteArrayOutputStream()) {
            if (length > Integer.MAX_VALUE) {
                throw new IOException("Size cannot be greater than Integer max value: " + length);
            }
            writeInputToOutPut(input, output, length);
            return output.toByteArray();
        }
    }

    /**
     * write input to output by length
     */
    private static void writeInputToOutPut(InputStream input, OutputStream output, long length) throws IOException {
        byte[] buffer = new byte[FILE_BUFFER_BLOCK];
        long hasReadLen = 0L;
        while (hasReadLen < length) {
            int readLen = (int) Math.min(length - hasReadLen, FILE_BUFFER_BLOCK);
            int len = input.read(buffer, 0, readLen);
            if (len != readLen) {
                throw new IOException("read" + hasReadLen + "bytes data less than " + length);
            }
            output.write(buffer, 0, len);
            hasReadLen += len;
        }
    }

    /**
     * Out put content to file.
     *
     * @param content Which content to out put
     * @param output  File to write
     * @throws IOException Write failed
     */
    public static void write(byte[] content, File output) throws IOException {
        if (output.exists() && !output.canWrite()) {
            CustomException.throwException(ERROR.WRITE_FILE_ERROR, SignToolErrMsg.FILE_WRITE_FAILED
                    .toString("No permission write to file " + output));
        }
        try (FileOutputStream out = new FileOutputStream(output)) {
            for (byte con : content) {
                out.write(con);
            }
        }
    }

    /**
     * Write data in file to output stream
     *
     * @param inFile input file path.
     * @param out output file path.
     * @param offset file read offset
     * @param size file read size
     * @return true, if write successfully.
     */
    public static boolean appendWriteFileByOffsetToFile(String inFile, FileOutputStream out, long offset, long size) {
        File inputFile = new File(inFile);
        try (FileInputStream fis = new FileInputStream(inputFile)) {
            fis.skip(offset);
            writeInputToOutPut(fis, out, size);
            return true;
        } catch (FileNotFoundException e) {
            LOGGER.error("Failed to get input stream object.");
        } catch (IOException e) {
            LOGGER.error("Failed to read or write data.");
        }
        return false;
    }

    /**
     * Check file exist or not.
     *
     * @param filePath File path
     * @return Is file exist
     */
    public static boolean isFileExist(String filePath) {
        return new File(filePath).exists();
    }

    /**
     * Throw runtime exception if not allowed file type.
     *
     * @param filePath file path
     * @param types    Such as "txt" "json" "mp3"
     */
    public static void validFileType(String filePath, String... types) {
        String suffix = getSuffix(filePath);
        ValidateUtils.throwIfNotMatches(!StringUtils.isEmpty(suffix),
                ERROR.NOT_SUPPORT_ERROR, SignToolErrMsg.NOT_SUPPORT_FILE.toString(filePath));
        boolean isMatches = false;
        for (String type : types) {
            if (StringUtils.isEmpty(type)) {
                continue;
            }
            if (type.equalsIgnoreCase(suffix)) {
                isMatches = true;
                break;
            }
        }
        ValidateUtils.throwIfNotMatches(isMatches,
                ERROR.NOT_SUPPORT_ERROR, SignToolErrMsg.NOT_SUPPORT_FILE.toString(filePath));
    }

    /**
     * Get suffix of file.
     *
     * @param filePath file path
     * @return file suffix. Such as "txt" "json" "p12"
     */
    public static String getSuffix(String filePath) {
        if (StringUtils.isEmpty(filePath)) {
            return "";
        }
        File file = new File(filePath);
        String fileName = file.getName();
        String[] temps = fileName.split("\\.");
        if (temps.length < SPLIT_LENGTH) {
            return "";
        }
        return temps[temps.length - 1];
    }

    /**
     * Write data in file to output stream
     *
     * @param file input file path.
     * @param dos output stream.
     * @return true, if write successfully.
     */
    public static boolean writeFileToDos(String file, DataOutputStream dos) {
        if (dos == null) {
            return false;
        }
        File src = new File(file);
        try (FileInputStream fileStream = new FileInputStream(src)) {
            int temp;
            byte[] buf = new byte[FILE_BUFFER_BLOCK];
            while ((temp = fileStream.read(buf)) > 0) {
                dos.write(buf, 0, temp);
            }
            return true;
        } catch (FileNotFoundException e) {
            LOGGER.error("Failed to get input stream object.");
        } catch (IOException e) {
            LOGGER.error("Failed to read or write data.");
        }
        return false;
    }

    /**
     * Write byte array to output stream
     *
     * @param data byte array
     * @param dos output stream
     * @return true, if write successfully.
     */
    public static boolean writeByteToDos(byte[] data, DataOutputStream dos) {
        if (data == null) {
            return true;
        }
        try {
            dos.write(data);
        } catch (IOException e) {
            LOGGER.error("Failed to write data to output stream.");
            return false;
        }
        return true;
    }

    /**
     * Get length of file.
     *
     * @param filePath input file path.
     * @return long value of file length.
     */
    public static long getFileLen(String filePath) {
        File file = new File(filePath);
        if (file.exists() && file.isFile()) {
            return file.length();
        }
        return -1;
    }

    /**
     * Write byte array data to output file.
     *
     * @param bytes byte array data.
     * @param outFile output file path.
     * @return true, if write successfully.
     */
    public static boolean writeByteToOutFile(byte[] bytes, String outFile) {
        try (OutputStream ops = new FileOutputStream(outFile, true)) {
            return writeByteToOutFile(bytes, ops);
        } catch (FileNotFoundException e) {
            LOGGER.error("Failed to get output stream object, outfile: " + outFile);
        } catch (IOException e) {
            LOGGER.error("Failed to write data to ops, outfile: " + outFile);
        }
        return false;
    }

    /**
     * Write byte array data to output file.
     *
     * @param bytes byte array data.
     * @param outFile output file path.
     * @return true, if write successfully.
     */
    public static boolean writeByteToOutFile(byte[] bytes, OutputStream outFile) {
        try {
            outFile.write(bytes, 0, bytes.length);
            outFile.flush();
            return true;
        } catch (FileNotFoundException e) {
            LOGGER.error("Failed to get output stream object, outfile: " + outFile);
        } catch (IOException e) {
            LOGGER.error("Failed to write data to ops, outfile: " + outFile);
        }
        return false;
    }

    /**
     * Check input file is valid.
     *
     * @param file input file.
     * @throws IOException file is a directory or can't be read.
     */
    public static void isValidFile(File file) throws IOException {
        if (!file.exists()) {
            throw new FileNotFoundException("File '" + file + "' does not exist");
        }

        if (file.isDirectory()) {
            throw new IOException("File '" + file + "' exists but is a directory");
        }

        if (!file.canRead()) {
            throw new IOException("File '" + file + "' cannot be read");
        }
    }

    /**
     * Open an input stream of input file safely.
     *
     * @param file input file.
     * @return an input stream of input file
     * @throws IOException file is a directory or can't be read.
     */
    public static FileInputStream openInputStream(File file) throws IOException {
        isValidFile(file);
        return new FileInputStream(file);
    }

    private static byte[] toByteArray(final InputStream input, final int size) throws IOException {
        if (size < 0) {
            throw new IllegalArgumentException("Size must be equal or greater than zero: " + size);
        }

        if (size == 0) {
            return EMPTY_BYTE_ARRAY;
        }

        final byte[] data = new byte[size];
        int offset = 0;
        int read;

        while (offset < size && (read = input.read(data, offset, size - offset)) != FILE_END) {
            offset += read;
        }

        if (offset != size) {
            throw new IOException("Unexpected read size. current: " + offset + ", expected: " + size);
        }

        return data;
    }

    /**
     * Read file to byte array.
     *
     * @param file input file.
     * @return byte array of file-content.
     * @throws IOException fill exception
     */
    public static byte[] readFileToByteArray(File file) throws IOException {
        try (InputStream in = openInputStream(file)) {
            final long fileLength = file.length();
            if (fileLength > Integer.MAX_VALUE) {
                throw new IllegalArgumentException("Size cannot be greater than Integer max value: " + fileLength);
            }
            return toByteArray(in, (int) fileLength);
        }
    }

    /**
     * Write a string to file
     *
     * @param source String will be written.
     * @param filePath path to write file.
     * @param charset a charset
     * @throws IOException write error.
     */
    public static void writeStringToFile(String source, String filePath, Charset charset) throws IOException {
        try (BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(
                new FileOutputStream(filePath), charset))) {
            writer.write(source);
            writer.flush();
        }
    }

    /**
     * Delete a file quietly
     *
     * @param file the file to delete
     */
    public static void deleteFile(File file) {
        if (file != null && file.isFile()) {
            try {
                Files.delete(file.toPath());
            } catch (IOException e) {
                LOGGER.warn("delete file '{}' error, error message: {}", file, e.getMessage());
            }
        }
    }

    /**
     * regex filename
     *
     * @param name filename
     * @return boolean
     */
    public static boolean isRunnableFile(String name) {
        if (StringUtils.isEmpty(name)) {
            return false;
        }
        if (name.endsWith(NATIVE_LIB_AN_SUFFIX) || name.endsWith(ABC_FILE_SUFFIX)) {
            return true;
        }
        if (name.startsWith(LIBS_PATH_PREFIX)) {
            return true;
        }
        return false;
    }
}
